### ISSUE TITLE
Budgie desktop 10.1 keyboard layout switching doesn't work

### ISSUE THREAD

**Issue by lexuv** [solus-project/budgie-desktop#269]
Budgie desktop 10.1 keyboard layout switching doesn't work ("Super+Space" works, but "Modificator key" (Ctrl-Shift) does not). And there's no keyboard layout indicator as well and it can't be enabled through panel applets. In full Gnome it works ok. Where to dig in?

**Comment by ikeydoherty** [comment]
`wm/keyboard.vala` is where the fault is at, we'll need to implement that there

For an indicator we'll likely need to expose some information between the WM and PanelManager (or just a generic dbus interface for an applet to know which layout the WM is using)

The WM already has dbus stuff there so could be a good place to start

**Comment by lexuv** [comment]
10.2.2 -- still is not fixed. :-(

**Comment by ikeydoherty** [comment]
@lexuv I never said it was, that's why this bug is still open

**Comment by improoviz** [comment]
Hi!
When it will be fixed?

**Comment by ikeydoherty** [comment]
As soon as it's done?

**Comment by improoviz** [comment]
unfortunately, i'm not a programmer.

**Comment by ikeydoherty** [comment]
I didn't ask if you were

**Comment by improoviz** [comment]
Have you a temporary solution?

**Comment by ikeydoherty** [comment]
No it needs implementing in the WM.

**Comment by ikeydoherty** [comment]
Alright I'm thick, what is a "Modificator key" ? Because I can't find this thing anywhere. We can change keyboard layouts already, so what is it we *can't* change ?

**Comment by lexuv** [comment]
Settings -> Keyboard -> Shortcuts -> Typing -> Modifiers-only switch to next source -> Ctrl-Shift. That doesn't work. Does not switch as it does in Gnome. And there no other way to set up, for example, Ctrl-Shift or Alt-Shift for swithing. There's no tray indicator of the current language as well. That's a big fail. Especially for non-english users.

**Comment by ikeydoherty** [comment]
Doesn't it just do exactly the same thing as switch input source?

**Comment by lexuv** [comment]
"Switch to nex input source" don't allow to set up Ctrl-Shift or Alt-Shift for switching. Try it.

**Comment by ikeydoherty** [comment]
OK but what's wrong with just using Super+Space ?

**Comment by ikeydoherty** [comment]
(This is the bit I'm confused about)

**Comment by lexuv** [comment]
It's not convinient. Most of ex windows users used to use Ctrl-Shift and Alt-Shift

**Comment by ikeydoherty** [comment]
Gotcha. But the behaviour it implements is the same, yes? That's just the bit I need to know, then I can hook it up very easily.

**Comment by lexuv** [comment]
Yes

**Comment by ikeydoherty** [comment]
Sweet! I'll do it after I've watched Netflix :)

**Comment by GeorgeTG** [comment]
Hi, any updates on this? As @lexuv said modifiers-only setting won't work. Trying to swatch to an other combination takes time to get used to and muscle memory makes this worse. Would be nice to get this to work!

**Comment by IlyaBizyaev** [comment]
Hello everyone!
Won't mind a keyboard layout indicator, too. That's what most environments provide by default and it's one of the reasons I'm not yet using Budgie as my default DE.

**Comment by IlyaBizyaev** [comment]
Have a bit different problem. The keyboard layout is switching randomly - sometimes you have to press multiple times to make it switch. Same user account has no problems when working on Unity.

**Comment by ikeydoherty** [comment]
This stuff is still WIP sorry

**Comment by IlyaBizyaev** [comment]
May be connected: https://github.com/solus-project/budgie-desktop/issues/419
By the way, layout switching seems to be not random. It can be divided into 3 issues:
1) Impossible to set Alt+Shift as a modificator
2) Combinations are changed like that: RU -> EN -> EN
3) The keyboard LED (in the same case): ON -> ON -> OFF.

**Comment by IlyaBizyaev** [comment]
Here's my **~/.inputrc**:
```
# im-config(8) generated on Fri, 22 Apr 2016 17:32:47 +0300
run_im xim
# im-config signature: <long code here>  -
```
Maybe that helps.

**Comment by Goomb** [comment]
Hi, I had similar issue. There were english and russian keyboard layout and english language set as main. When I tried to switch layout, I have cycle RU -> EN -> EN. But when I added russian locale, set RU as main in system and delete RU from Input Source, it's all ok now. Only RU  -> EN when press Win-Space. Maybe it'ii be useful for somebody.

**Comment by ikeydoherty** [comment]
The extranous layout was fixed in git :)

**Comment by Goomb** [comment]
Oh, sorry. I missed this. Maybe last stable Budgie-desktop for Arch not updated yet.

**Comment by ikeydoherty** [comment]
Yeah Arch is using the 10.2.6 release, the changes will come about in 10.2.7, as they're in between right now. They're not in Solus either, don't worry ^^

**Comment by ikeydoherty** [comment]
So yeah closing as this was fixed in 10.2.7 + 10.2.8

**Comment by yellowmegaman** [comment]
Got same thing in `10.2.9-2`

**Comment by mhsabbagh** [comment]
This is still not fixed, neither in Solus nor Ubuntu Budgie. It's really a blocking bug.

**Comment by ikeydoherty** [comment]
How is this a blocking bug? You hit super+space and it changes the keyboard layout. The base functionality  *is there*. There is also a keyboard layout indicator. What isn't there is some fuzzy concept that everyone has thus far failed to explain to me, about using a different shortcut with keys unsupported by the underlying stack (Two modifiers in conjunction).

**Comment by ikeydoherty** [comment]
TL;DR Explain it to me like I'm 5 :P Why keyboard layout switching is considered to be broken when I know it to be working

**Comment by YurySolovyov** [comment]
Which component do you mean by underlying stack? Say, Ubuntu in general does that perfectly well with Ctrl+Shift and Alt+Shift

**Comment by ikeydoherty** [comment]
GNOME, obviously I don't use Ubuntu.

That said I've not even touched this stuff in since-forever. So it may well be restored again with 3.24

**Comment by mhsabbagh** [comment]
Even GNOME supports switching using Alt + Shift (I have been using it for years on GNOME). The problem is that I've been using this keyboard shortcut for all of my life, on Linux and Windows. The fact that I have to use another keyboard shortcut on a specific desktop environment (or distro) just because it doesn't support it really makes it feel like broken :P

**Comment by mhsabbagh** [comment]
This is from GNOME Tweak Tool:

![Uploading Screenshot from 2017-04-20 06-14-16.png…]()

**Comment by mhsabbagh** [comment]
This is from GNOME Tweak Tool:

![screenshot from 2017-04-20 06-14-16](https://cloud.githubusercontent.com/assets/1293003/25212021/b9ce2874-2590-11e7-8d1f-2aa98c16b9f4.png)

**Comment by ikeydoherty** [comment]
>  just because it doesn't support it really makes it feel like broke

Budgie is **based** on GNOME. It isn't GNOME Shell. GNOME Tweak Tool is for GNOME shell.
It's **certainly** not broken.

If you put a feature request in for **Budgie 11** for configurable shortcuts, sure. But not on 10.3 with the hybrid stack, I've told people repeatedly there will be no new features landing. 10.x is done. Also note that GNOME Tweak Tool will definitely not be supported for Budgie 11.

**Comment by mhsabbagh** [comment]
Sure, no matter that way or which version, offering such option to switch layout shortcuts is something needed by a lot of people here.

Should I open a new bug and mark it with "feature request"?

**Comment by ahmadabdel-halim** [comment]
I want to use another shortcut to switch between languages really. In my case, I want to use alt + shift, but I can't do that.

**Comment by lexuv** [comment]
You were asked to fix this two (!) yeras ago. And still haven't done it. This is the feature that all the non-english-spoken users use all the time. Where's user-friendly support?

**Comment by fossfreedom** [comment]
@lexuv that really is unfair - the code is open.  If the community wants this then they should step up and supply the code-patch.  support is two way.

**Comment by YurySolovyov** [comment]
Even though I've run into this too, I want to thank @ikeydoherty for his patience on this.
I'd be happy to follow v11 issue if that's the way forward.

**Comment by ikeydoherty** [comment]
@lexuv dude you're way out of line there. I have a lot of projects going on. Truth remains that in the 4 years that Budgie has been around, I'm still the #1 contributor and well over 90% of the code is mine.
People can step up, but many have chosen not to. If Budgie is as popular as people say, then perhaps patches could accompany that good will.

> You were asked to fix this two (!) years ago

Let's get one thing straight here. I'm not obliged to do anything anyone tells me to. I do it if i can, and in line with the roadmap and current activities I have going on. You're asking me to be "user friendly" yet in the same breath feel you have some power over me? Not fucking cool, lad.

We all know that the 10.3 codebase leaves a lot to be desired. At this point adding any new features to it equates to flogging a dead horse. I want to move on and start the 11.x codebase so we can get rid of *all of our issues*, once and for all. If you continue to block me on this, how do we expect 11.x to ever see the light of day?

I've been presented with an ill-described use-case that I know nothing about, which equates to "GNOME does this so should you" - despite knowing that 11.x *will not* be GNOME based. I have at least one more maintenance release going out, which will add the ability to mute the notifications. If someone lands a patch before then (or hell, a decent explanation and pointers) then maybe it can go in.

Not when you're talking to me like that, though.

**Comment by ikeydoherty** [comment]
@mhsabbagh yep lets do a new feature request for it, with the details I need to know (Remember - explain like I'm five, I don't use these features) and we'll take it from there and get this back on track.

I hereby bury this task.

**Comment by YurySolovyov** [comment]
@mhsabbagh make sure to mention this one so everyone can follow.

**Comment by Split7fire** [comment]
If someone still facing this issue...
I managed to workaround this problem by the help from dconf-editor:
1. Find /org/gnome/desktop/wm/keybindings/switch-input-source in dconf-editor
2. Change default value to ['<Alt>Shift_L', '<Alt>Shift_R']
3. Profit.

P.S. Note that now you can change layouts by presssing Alt and then Shift, either right or left.

**Comment by spijet** [comment]
Unfortunately, the solution proposed by @Split7fire doesn't seem to work on current Solus (fresh Solus Budgie, installed yesterday). If I set XKB options via `localectl` or via direct editing of XOrg config files, I can use the second layout in LightDM, but not in Budgie session.

I also noticed a strange behavior: I use 2 keyboard layouts (US English and Russian) with a `grp_led:scroll` XKB option (which makes "Scroll Lock" keyboard LED light up when Russian layout is active). When I try to change layouts in LightDM, everything works OK and the LED lights up until I change the layout back to US. In Budgie though, if I try to change the layout to Russian, the LED only blinks (that is, lights up for a really short period of time) once.

To add to it all: I observe the same behavior in current Arch Linux with Budgie installed. I'll leave the package details below, in case if that helps to clarify some things.
```
Name            : budgie-desktop
Version         : 10.4+8+gff09a510-1
Packager        : Balló György <ballogyor+arch@gmail.com>
Build Date      : Sun 15 Oct 2017 10:06:18 PM +08
Install Date    : Fri 08 Dec 2017 07:03:34 PM +08
```

**Comment by L11R** [comment]
It was 2018. But we still cannot change f**king layout 😑

**Comment by mhsabbagh** [comment]
Good good, I will give Solus 4 a try when it comes out then.

Cheers!

**Comment by savchenko91** [comment]
['<Shift>Alt_L'] - also works

**Comment by SerafimArts** [comment]
@savchenko91 then that would be for sure
```
['<Alt>Shift_L', '<Alt>Shift_R', '<Shift>Alt_L', '<Shift>Alt_R']
```

=)

**Comment by covrom** [comment]
['<Control>Shift_L', '<Control>Shift_R', '<Shift>Control_L', '<Shift>Control_R']

**Comment by scythargon** [comment]
![image](https://user-images.githubusercontent.com/1755870/69530244-c1371880-0f82-11ea-9324-8c22a4d5de91.png)
Still does not switch for me:(
I did for some time but after reboot it stopped. I've tried to reset values back and forth, no luck.

**Comment by notio-ca** [comment]
Did everything on Ubuntu, still not working... I am switching from input source with ATL-CTRL since 20 years... 5 years in Linux, even i3wm, Gnome, Xfce I can do this, please, I just can't hack my brain, well, Budgie don't deserve that, this is suppose to be a lightweight **well polished** desktop
